src/esp/assets/GenericInstanceMeshData.cpp

```diff
@@ -94,33 +105,183 @@ Cr::Containers::Optional<InstancePlyData> parsePly(
+
+        // derive semantic ID and color and region/room ID for culling
+        int regionID = 0;
+        int semanticID = 0;
+        if (ssdColorToIDAndRegionIter !=
+            tmpColorMapToSSDidAndRegionIndex.end()) {
+          // color is found in ssd mapping, so is legal color
+          semanticID = ssdColorToIDAndRegionIter->second.first;
+          regionID = ssdColorToIDAndRegionIter->second.second;
+        } else {
+          // color is not found in ssd mapping, so not legal color
+          // check if we've assigned a semantic ID to this color before, and if
+          // not do so
+          auto nonSSDClrRes =
+              nonSSDVertColors.insert({meshColorInt, nonSSDObjID});
+          if (nonSSDClrRes.second) {
+            // inserted, so increment nonSSDObjID
+            ++nonSSDObjID;
+          }
+          // map holds that color's nonSSDObjID
+          semanticID = nonSSDClrRes.first->second;
+          regionID = maxRegion;
+        }
+        // semantic ID for vertex
+        data.objectIds[vertIdx] = semanticID;
+        // color for
+        tmpDestSSDColorMap[semanticID] = meshColor;
+        // partition Ids for each vertex, for multi-mesh construction.
+        data.partitionIds[vertIdx] = regionID;
+      }
+
+      // color map with first nonSSDObjID-1 elements in proper order to match
+      // provided semantic IDs (so that ID is IDX of semantic color in map) and
+      // any overflow colors uniquely map 1-to-1 to an unmapped semantic ID as
+      // their index.
+      Cr::Containers::Array<Mn::Vector3ub> tmpDestSSDColorVec{
+          Mn::NoInit, tmpDestSSDColorMap.size()};
+      for (const auto& elem : tmpDestSSDColorMap) {
+        tmpDestSSDColorVec[elem.first] = elem.second;
+      }
+
+      // colors need to be in the appropriate order
+      colorMapToUse.resize(tmpDestSSDColorVec.size());
+      Cr::Utility::copy(tmpDestSSDColorVec,
+                        Cr::Containers::arrayView(colorMapToUse));
+
+    } else {
+      // no remapping provided by SSD so just use what is synthesized
+      Cr::Containers::Array<Mn::Color3ub> colorsThatBecomeTheColorMap{
+          Mn::NoInit, meshColors.size()};
+      // copy the meshColors into colorsThatBecomeTheColorMap
+      Cr::Utility::copy(meshColors, colorsThatBecomeTheColorMap);
+
+      // removeDuplicatesInPlace returns array of unique ids for colors, and
+      std::pair<Cr::Containers::Array<Mn::UnsignedInt>, std::size_t> out =
+          Mn::MeshTools::removeDuplicatesInPlace(
+              Cr::Containers::arrayCast<2, char>(
+                  stridedArrayView(colorsThatBecomeTheColorMap)));
+
+      objectIds = std::move(out.first);
+      const Corrade::Containers::ArrayView<Magnum::Vector3ub>
+          clrsListFromVerts = Cr::Containers::arrayCast<Mn::Vector3ub>(
+              colorsThatBecomeTheColorMap.prefix(out.second));
+      colorMapToUse.assign(clrsListFromVerts.begin(), clrsListFromVerts.end());
```

Comment:
Won't just this work? IIRC `assign()` doesn't care what input type it is as long as it's implicitly convertible.

```suggestion
      colorMapToUse.assign(colorsThatBecomeTheColorMap.begin(), colorsThatBecomeTheColorMap.begin() + out.second);
```